Description: Align StSpinnerContent Design and Animation to YaruCircularProgressIndicator
Author: Vincent Renzo Quilon
Forwarded: no, ubuntu only.
Last-Update: 2025-06-25
---
This patch header follows DEP-3: http://dep.debian.net/deps/dep3/
--- a/src/st/st-spinner-content.c
+++ b/src/st/st-spinner-content.c
@@ -23,21 +23,32 @@
 
 #define MIN_RADIUS 8
 #define NAT_RADIUS 48
-#define SMALL_WIDTH 2.5
-#define LARGE_WIDTH 12
-#define SPIN_DURATION_MS 1200
+/*
+ * AdwSpinnerPaintable's MAX_RADIUS has similar function to NAT_RADIUS
+ * but NAT_RADIUS = 48 while MAX_RADIUS = 32.
+ * to calculate new LARGE_WIDTH, we apply LERP
+ * where t = (NAT_RADIUS - 8) / (MAX_RADIUS - 8) = 1.67
+ * AdwSpinnerPaintable's LARGE_WIDTH = 6
+ * new LARGE_WIDTH = LERP (2, 6, 1.67) = 8.67
+ */
+#define LARGE_WIDTH 8.67
+#define SMALL_WIDTH 2
 #define START_ANGLE (G_PI * 0.35)
 #define CIRCLE_OPACITY 0.15
-#define MIN_ARC_LENGTH (G_PI * 0.015)
-#define MAX_ARC_LENGTH (G_PI * 0.9)
-#define IDLE_DISTANCE (G_PI * 0.9)
-#define OVERLAP_DISTANCE (G_PI * 0.7)
-#define EXTEND_DISTANCE (G_PI * 1.1)
-#define CONTRACT_DISTANCE (G_PI * 1.35)
-/* How many full cycles it takes for the spinner to loop. Should be:
- * (IDLE_DISTANCE + EXTEND_DISTANCE + CONTRACT_DISTANCE - OVERLAP_DISTANCE) * k,
- * where k is an integer */
-#define N_CYCLES 53
+#define SEGMENT_ANGLE (2 * G_PI / 3) // 120 degress per segment
+#define SPIN_DURATION_MS 8000
+#define N_CYCLES 6
+/*
+ * AdwSpinnerPaintable UPPER_MIN_GAP_ANGLE = 0.2 at MAX_RADIUS = 32
+ * LOWER_MIN_GAP_ANGLE =  0.1 at MIN_RADIUS = 8.
+ * note: MAX_RADIUS != NAT_RADIUS.
+ * to calculate the new UPPER_MIN_GAP_ANGLE, we apply LERP
+ * where t = (NAT_RADIUS - 8) / (MAX_RADIUS - 8) = 1.67
+ * new UPPER_MIN_GAP_ANGLE = LERP (0.1, 0.2, 1.67) = 0.267
+ */
+#define UPPER_MIN_GAP_ANGLE 0.267
+#define LOWER_MIN_GAP_ANGLE 0.1
+#define MAX_GAP_ANGLE 0.5
 
 /**
  * StSpinnerContent:
@@ -127,9 +138,9 @@ st_spinner_content_set_actor (StSpinnerC
   if (spinner->actor)
     {
       spinner->timeline = clutter_timeline_new_for_actor (actor,
-                                                          SPIN_DURATION_MS * N_CYCLES);
+                                                          SPIN_DURATION_MS);
       clutter_timeline_set_repeat_count (spinner->timeline, -1);
-      clutter_timeline_set_progress_mode (spinner->timeline, CLUTTER_LINEAR);
+      clutter_timeline_set_progress_mode (spinner->timeline, CLUTTER_EASE_IN_OUT_SINE);
 
       g_signal_connect (spinner->timeline, "new-frame", G_CALLBACK (new_frame_cb), spinner);
 
@@ -200,46 +211,6 @@ normalize_angle (double angle)
   return angle;
 }
 
-static inline double
-ease_in_out_sine (double t)
-{
-  return -0.5 * (cos (M_PI * t) - 1);
-}
-
-static double
-get_arc_start (double angle)
-{
-  double l = IDLE_DISTANCE + EXTEND_DISTANCE + CONTRACT_DISTANCE - OVERLAP_DISTANCE;
-  double t;
-
-  angle = fmod (angle, l);
-
-  if (angle > EXTEND_DISTANCE)
-    t = 1;
-  else
-    t = ease_in_out_sine (angle / EXTEND_DISTANCE);
-
-  return LERP (MIN_ARC_LENGTH, MAX_ARC_LENGTH, t) - angle * MAX_ARC_LENGTH / l;
-}
-
-static double
-get_arc_end (double angle)
-{
-  double l = IDLE_DISTANCE + EXTEND_DISTANCE + CONTRACT_DISTANCE - OVERLAP_DISTANCE;
-  double t;
-
-  angle = fmod (angle, l);
-
-  if (angle < EXTEND_DISTANCE - OVERLAP_DISTANCE)
-    t = 0;
-  else if (angle > l - IDLE_DISTANCE)
-    t = 1;
-  else
-    t = ease_in_out_sine ((angle - EXTEND_DISTANCE + OVERLAP_DISTANCE) / CONTRACT_DISTANCE);
-
-  return LERP (0, MAX_ARC_LENGTH - MIN_ARC_LENGTH, t) - angle * MAX_ARC_LENGTH / l;
-}
-
 static void
 st_spinner_content_draw_spinner (StSpinnerContent *spinner,
                                  cairo_t          *cr,
@@ -247,8 +218,10 @@ st_spinner_content_draw_spinner (StSpinn
                                  int               height)
 {
   CoglColor color;
-  double radius, line_width;
+  double radius, radius_t, line_width;
   double progress, start_angle, end_angle;
+  float gap_angle, min_gap_angle, gap_angle_t_actual, gap_angle_t;
+  uint8_t segment_counter;
 
   g_assert (spinner->actor);
 
@@ -263,6 +236,7 @@ st_spinner_content_draw_spinner (StSpinn
     }
 
   radius = MIN (floorf (MIN (width, height) / 2), NAT_RADIUS);
+  radius_t = INVERSE_LERP (MIN_RADIUS, NAT_RADIUS, radius);
   line_width = LERP (SMALL_WIDTH, LARGE_WIDTH,
                      INVERSE_LERP (MIN_RADIUS, NAT_RADIUS, radius));
   radius -= roundf (line_width / 2.);
@@ -273,46 +247,46 @@ st_spinner_content_draw_spinner (StSpinn
   cairo_translate (cr, roundf (width / 2), roundf(height / 2));
   cairo_set_line_width (cr, line_width);
 
-  /* Circle */
-
-  cairo_save (cr);
-
-  cairo_set_source_rgba (cr,
-                         color.red / 255.,
-                         color.green / 255.,
-                         color.blue / 255.,
-                         color.alpha / 255. * CIRCLE_OPACITY);
-  cairo_arc (cr, 0, 0, radius, 0, 2 * M_PI);
-  cairo_stroke (cr);
-
-  cairo_restore (cr);
-
   /* Moving part */
 
-  cairo_save (cr);
+  segment_counter = 0;
 
   if (spinner->timeline)
     progress = clutter_timeline_get_progress (spinner->timeline) * N_CYCLES * M_PI * 2;
   else
-    progress = EXTEND_DISTANCE - OVERLAP_DISTANCE / 2;
+    progress = START_ANGLE;
 
-  start_angle = progress + get_arc_start (progress) + START_ANGLE;
-  end_angle = progress + get_arc_end (progress) + START_ANGLE;
 
-  start_angle = normalize_angle (start_angle);
-  end_angle = normalize_angle (end_angle);
+  min_gap_angle = LERP (UPPER_MIN_GAP_ANGLE, LOWER_MIN_GAP_ANGLE, radius_t);
+  gap_angle_t_actual = INVERSE_LERP (0, 2 * M_PI * N_CYCLES, progress);
+  gap_angle_t = 1 - fabs (2 * gap_angle_t_actual - 1); // pinpong behavior: 0->1 is transformed to 0->1->0
+  gap_angle = LERP (min_gap_angle, MAX_GAP_ANGLE, gap_angle_t);
 
-  cairo_set_line_cap (cr, CAIRO_LINE_CAP_ROUND);
-  cairo_set_source_rgba (cr,
-                         color.red / 255.,
-                         color.green / 255.,
-                         color.blue / 255.,
-                         color.alpha / 255.);
+  end_angle = progress + START_ANGLE - SEGMENT_ANGLE;
+  start_angle = progress + START_ANGLE - gap_angle;
 
-  cairo_arc (cr, 0, 0, radius, end_angle, start_angle);
-  cairo_stroke (cr);
+  do {
+    cairo_save (cr);
 
-  cairo_restore (cr);
+    start_angle = normalize_angle (start_angle);
+    end_angle = normalize_angle (end_angle);
+
+    cairo_set_line_cap (cr, CAIRO_LINE_CAP_BUTT);
+    cairo_set_source_rgba (cr,
+                           color.red / 255.,
+                           color.green / 255.,
+                           color.blue / 255.,
+                           color.alpha / 255.);
+
+    cairo_arc (cr, 0, 0, radius, end_angle, start_angle);
+    cairo_stroke (cr);
+
+    segment_counter++;
+    start_angle += SEGMENT_ANGLE;
+    end_angle += SEGMENT_ANGLE;
+    cairo_restore (cr);
+
+  } while (segment_counter < 3);
 }
 
 static void
@@ -446,3 +420,4 @@ st_spinner_content_new (void)
 {
   return g_object_new (ST_TYPE_SPINNER_CONTENT, NULL);
 }
+
